### ISSUE TITLE
[1.16.5] Custom ItemEntity Implementations lead to duplicated items

### ISSUE THREAD

**Issue by HellFirePvP** [MinecraftForge/MinecraftForge#7785]
**Minecraft Version:** 1.16.5

**Forge Version:** 36.1.23 (Any 1.16 forge will do)

**Steps to Reproduce:**
 1. Create Item which overrides `hasCustomEntity` to return true and `createEntity` to return a custom Entity for the item when dropped (can also just be another/new ItemEntity instance)
 2. /give the Item in question ingame, while having inventory space for the item
 3. The item is given to the player AND a 2nd copy is dropped

**Description of issue:**

The ItemStack is given to the player in the /give command via `serverplayerentity.inventory.addItemStackToInventory(itemstack);` (GiveCommand L39) and is then in the player's inventory if the player has space for it (see Step2).
However, for minecraft's /give "flavour" a fake item is to be created, via ServerPlayerEntity#dropItem(ItemStack, boolean) (GiveCommand L42), which ends up calling ForgeHooks#onPlayerTossEvent.
Since the custom item entity implementation relies on the EntityJoinWorldEvent to replace the dropped ItemEntity, in ForgeHooks#onPlayerTossEvent at L477 World#addEntity(Entity), the given ItemEntity is ultimately replaced in the JoinWorld with the actual real entity created by Item#createEntity's custom entity implementation.
This results in the return statement L478, event.getEntityItem, returning a item entity which is by then already dead/removed (see ForgeInternalHandler#onEntityJoinWorld L65 Entity#remove) since it was replaced by the "custom" item entity. The now old/removed entity gets returned and turned into a fake-item in GiveCommand#giveItem L44 ItemEntity#makeFakeItem.
Unfortunately, the item turned into a fake item there is already dead and to be replaced with the custom item entity. However, Item#createEntity has no information of any of this happening since the Entity passed in is a completely valid item entity at that point and the ItemStack given is valid as well since the ItemStack passed into dropItem in GiveCommand L42 is set to count 1 the line before to create a "correct" fake item entity. There is no way to detect this happening from the custom item entity implementation side.

Unfortunately, this kind of behavior seems to exist for several other methods where mods attempt to collect items dropped from chest breaking or otherwise due to there not being a clean way to detect this from occurring.
Even more so since the custom itementity is only dropped the next tick instead of being created right away, see ForgeInternalHandler#onEntityJoinWorld L68 ThreadTaskExecutor#enqueue

**Comment by TheCurle** [comment]
This seems to be an issue with your implementation of the ItemEntity.
The pickupDelay is normally supposed to be set to it's max value, to prevent this exact bug. 

See here: https://github.com/MinecraftForge/MinecraftForge/issues/7481#issuecomment-749100352

**Comment by HellFirePvP** [comment]
I'm sorry, but this issue is not the issue you're referencing; the underlying problem is different and doesn't have anything to do with the pickupDelay.

**Comment by diesieben07** [comment]
You are correct in your assessment of this issue. But I don't see a way to fix this other than to patch every single place that creates a new `ItemEntity` instance.

**Comment by HellFirePvP** [comment]
Though it is not a clean solution, pupnewfster suggested changing my implementation and keeping the old item-entity instance around and double-checking what happens to it later on and thus finding out if it's turned into a fake-item for /give purposes or not, to at least take care of that incident.

**Comment by Technici4n** [comment]
This might be worth looking into before the 1.18 RB in case a breaking change is necessary? There was also an issue with mekanism cardboard boxes failing to cancel custom item entities spawns, not sure if it's related.

Edit: I was referring to mekanism/Mekanism#7053.
